rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,802 @@
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    async fn report_data(
+        &self,
+        id: u32,
+        node_id: u64,
+        rx: &[u8],
+        tx: &mut [u8],
+        exchange: &mut Exchange<'_>,
+    ) -> Result<bool, Error>
+    where
+        T: DataModelHandler,
+    {
+        let mut wb = WriteBuf::new(tx);
+
+        let req = SubscribeReq::from_tlv(&get_root_node_struct(rx)?)?;
+        let req = ReportDataReq::Subscribe(&req);
+
+        let metadata = self.handler.lock().await;
 
-                    driver.complete(req).await?;
+        let accessor = exchange.accessor()?;
+
+        {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
+
+            loop {
+                let more_chunks = req
+                    .respond(&self.handler, Some(id), &mut attrs, &mut wb, false)
+                    .await?;
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                if !Self::recv_status(exchange).await? {
+                    info!("Subscription {node_id:x}::{id} removed during reporting");
+                    return Ok(false);
                 }
-                Interaction::Subscribe {
-                    req,
-                    ref mut driver,
-                } => {
-                    let accessor = driver.accessor()?;
-
-                    'outer: for item in metadata.node().subscribing_read(req, None, &accessor) {
-                        while !AttrDataEncoder::handle_read(&item, &self.0, &mut driver.writer()?)
-                            .await?
-                        {
-                            if !driver.send_chunk(req).await? {
-                                break 'outer;
-                            }
-                        }
-                    }
 
-                    driver.complete(req).await?;
+                if !more_chunks {
+                    break;
                 }
             }
         }
 
+        Ok(true)
+    }
+
+    async fn rx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        let buffer = self.tx_buffer(exchange).await?;
+
+        if let Some(mut buffer) = buffer {
+            let rx = exchange.rx()?;
+
+            buffer.clear();
+            buffer.extend_from_slice(rx.payload()).unwrap();
+
+            exchange.rx_done()?;
+
+            Ok(Some(buffer))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn tx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        if let Some(mut buffer) = self.buffers.get().await {
+            buffer.resize_default(MAX_EXCHANGE_TX_BUF_SIZE).unwrap();
+
+            Ok(Some(buffer))
+        } else {
+            Self::send_status(exchange, IMStatusCode::Busy).await?;
+
+            Ok(None)
+        }
+    }
+
+    async fn recv_status(exchange: &mut Exchange<'_>) -> Result<bool, Error> {
```

Comment:
naming wise, I would expect a `receive_status` to return a status rather than a boolean.

I believe this should either return the status or maybe be named `recv_success`. However then if I think of this as success, this seems to be able to return `true/success`, `false/not-success`, Error ... and then I wonder `what is the difference between error and not a success`

How about having a separate enum then like `ReceivedStatus::Success, ReceivedStatus::Other(u8)` that gets returned here. That way we can check for Success, also do not lose the status on failure and the return is not a bool where false is not as clear?

Reply:
> naming wise, I would expect a `receive_status` to return a status rather than a boolean.
> 
> I believe this should either return the status or maybe be named `recv_success`. However then if I think of this as success, this seems to be able to return `true/success`, `false/not-success`, Error ... and then I wonder `what is the difference between error and not a success`

Returning an error is any other unexpected condition. Which could be anything. Returning a failure status however is _not_ an unexpected error and has a well-defined semantics. As in the other party telling you to abort the current transaction. Rare, but might happen and is defined in the spec.

Sure, we can emulate returning a non-successful status with error codes (possibly, by extending the `ErrorCode` enum) but I'm not sure we are winning much here.

> How about having a separate enum then like `ReceivedStatus::Success, ReceivedStatus::Other(u8)` that gets returned here. That way we can check for Success, also do not lose the status on failure and the return is not a bool where false is not as clear?

Sure we can, but if we do this, then what? We still have to handle the `ReceivedStatus::Other(u8)` case, and handling it is always done the same way:
```
warn!(
                "Got status response {:?}, aborting interaction",
                resp.status
            );

            drop(rx);
            exchange.acknowledge().await?;

            Ok(false)
```

Since this handling is the same for all transactions, then we ideally have to keep it inside the `recv_status` method. But if we do that, then what is the point of returning `ReceivedStatus::Other(u8)` in the first place? The exact returned status code is logged already, so code up the stack does not need it?

I'll rather rename `recv_status` to `recv_status_success` instead.